Automatic docker push after build

uilianries opened the pull request:
Hi!

You will no longer need to execute the same push routine any more! I added some new scripts to build, test and push. So, test.sh, build.sh and run.sh are no longer needed.

Before to accept these changes, you need to configure DOCKER_PASSWORD in Travis, if you want upload some image.

You can see all results here: https://travis-ci.org/uilianries/conan-docker-tools/builds/238598032 and here are the images: https://hub.docker.com/r/uilianries/  
My last build has failed, because it is using lasote as docker username ;)

Saludos!

lasote commented on README.md:
```diff
@@ -93,3 +93,45 @@ docker run -v/tmp/.conan:/home/conan/.conan lasote/conangcc63 bash -c "conan ins
+If you do not set any compiler version, the script will execute all supported versions.
+
+### Test
+The second stage runs the new image and builds ``gtest/1.8.0`` with libstdc++, for x86 and x86_64. ``Clang`` images use libc++ and libstdc++.
```
Specifying the setting `compiler` and `compiler.version` in the build to ensure that the current installed compiler is correct.

lasote commented on conan_docker_tools/application.py:
```diff
@@ -0,0 +1,71 @@
+        self.__process_compiler(GccCompiler(), gcc_versions)
+        self.__process_compiler(ClangCompiler(), clang_versions)
+
+    def __process_compiler(self, compiler, versions):
```
For private methods we use only one "_" underscore by convention, it allows the IDE etc to access the code.

lasote commented on conan_docker_tools/compiler.py:
```diff
@@ -0,0 +1,35 @@
+"""Static compiler information
+"""
+
+class Compiler(object):
```
NamedTuple("Compiler", "name, versions")

lasote commented on README.md:
````diff
@@ -93,3 +93,45 @@ docker run -v/tmp/.conan:/home/conan/.conan lasote/conangcc63 bash -c "conan ins
+Supported variables:
+
+```
+CONAN_GCC_VERSIONS     GCC versions to build, test and deploy
````
In which format? [4.6, 4.9] ? [46,49] Provide an example

lasote commented on README.md:
```diff
@@ -93,3 +93,45 @@ docker run -v/tmp/.conan:/home/conan/.conan lasote/conangcc63 bash -c "conan ins
+Build, Test and Deploy
+======================
+
+To create all default images, this project uses a bunch of script in conan_docker_tools. These scripts run as an application,
```
Introduce: The images are already built and uploaded to "lasote" dockerhub account, If you want to build your own images you can do it by...

lasote commented on conan_docker_tools/build.py:
```diff
@@ -0,0 +1,14 @@
+from image import Image
+
+
+class Build(object):
```
Replace this class with a simple function.

lasote commented on conan_docker_tools/compiler.py:
```diff
@@ -0,0 +1,35 @@
+    """Specilized builder for GCC
+    """
+    def __init__(self):
+        Compiler.__init__(self, "gcc", ["4.6", "4.8", "4.9", "5.2", "5.3", "5.4", "6.2", "6.3"])
```
The right way to call the parent constructor is:

super(GccCompiler, self).__init__(

lasote commented on conan_docker_tools/compiler.py:
```diff
@@ -0,0 +1,35 @@
+    """Specialized builder for Clang
+    """
+    def __init__(self):
+        Compiler.__init__(self, "clang", ["3.8", "3.9", "4.0"])
```
super(ClangCompiler, self).init(

lasote commented on conan_docker_tools/deploy.py:
```diff
@@ -0,0 +1,21 @@
+import sys
+
+
+class Deploy(object):
```
A class is not needed, just create a function

lasote commented on conan_docker_tools/test.py:
```diff
@@ -0,0 +1,49 @@
+        return TestGcc() if image.compiler == "gcc" else TestClang()
+
+
+class TestGcc(Test):
```
function test_gcc

lasote commented on conan_docker_tools/test.py:
```diff
@@ -0,0 +1,49 @@
+            raise
+
+
+class TestClang(TestGcc):
```
function test_clang

lasote commented on conan_docker_tools/image.py:
```diff
@@ -0,0 +1,30 @@
+from variables import Variables
+
+
+class Image(object):
```
Can be replaced with a NamedTuple and just a "get_variables" function

lasote commented:
Hi!! thanks for your help.

In general, the PR is ok, my main concerns are:

- In python we always try to avoid boilerplate classes, in Java, your PR would be perfect, but in python is very very verbose. We try to avoid the "bean" clases with private attributes and properties, if somone is going to access the property just keep them public at least there is a critical reason to not allow the write.
- Most of the classes can be replaced with tiny functions. 
- In the .travis.yml we need to introduce some check like TRAVIS_BRANCH == "master" to avoid upload images with a PR (anyone could override my images if I set the password). If don't just don't run the upload, but not fail, we need to check if the PR is ok.
- All the arguments you can pass to the program to generate the classes, maybe is oversized, as a principle we like to go step by step and only develop a feature when it's really needed. Do you think that many people are going to build the images by their own and use these parameters? I don't.
- For that reason, I think we should simplify the code, maybe keep the latest version and just introduce the possibility of upload the images (just with a new function).

Thanks for all your efforts! 👍

uilianries commented:
Thanks for the review and comments!

I'll rectify ASAP.

Regards.

uilianries commented:
Done! @lasote 

I simplified all to build.py.

I just removed DOCKER_USERNAME from Travis file to enable anyone to create PR without errors.

The conan-package-tools was a good project to take a look. I brought CONAN_STABLE_BRANCH_PATTERN from there.

Now, you will need to include DOCKER_USERNAME and DOCKER_PASSWORD in your Travis settings, but this is documented.

Have a nice weekend!

lasote commented on build.py:
```diff
@@ -0,0 +1,136 @@
+            return
+
+        logging.info("Login to Docker hub account")
+        result = subprocess.call(['sudo', 'docker', 'login', '-p', self.variables.docker_password, '-u', self.variables.docker_username])
```
Maybe avoid sudo and append the sudo in the .travis.yml?

lasote commented on build.py:
```diff
@@ -0,0 +1,136 @@
+            raise RuntimeError("Could not login username %s to Docker hub." % self.variables.docker_username)
+
+        logging.info("Upload Docker image %s to Docker hub." % image_name)
+        subprocess.check_call("sudo docker push %s" % image_name, shell=True)
```
Idem

lasote commented:
Now It looks very good! much more pythonic ;) Good job.
About the "sudo" in commands, maybe we should remove it and append it to the .travis / python run. For example in OSX is not needed to use sudo in docker commands.

uilianries commented:
Good point, I'm remove.

uilianries commented:
@lasote Done!

lasote commented:
Hi, I'm taking a look to this and finally see if we can improve the automatic creation of the images.

lasote commented:
I've simplified a couple things (no more default lasote and renamed image env vars, no more default images), added the conan_server image building:

https://github.com/conan-io/conan-docker-tools/pull/19

Let's see how it goes. All your commits are there, so you are the main contributor of this feature ;)
